File: templates/Uwp/Pages/InkSmartCanvas/ViewModels/InkSmartCanvasViewViewModel.cs
using Param_ItemNamespace.Services.Ink;
using Param_ItemNamespace.Services.Ink.UndoRedo;
using System.Linq;

namespace Param_ItemNamespace.ViewModels
{
    public class InkSmartCanvasViewViewModel : System.ComponentModel.INotifyPropertyChanged
    {
        private InkStrokesService _strokeService;
        private InkLassoSelectionService _lassoSelectionService;
        private InkNodeSelectionService _nodeSelectionService;
        private InkPointerDeviceService _pointerDeviceService;
        private InkUndoRedoService _undoRedoService;
        private InkTransformService _transformService;
        private InkFileService _fileService;

        private RelayCommand undoCommand;
        private RelayCommand redoCommand;
        private RelayCommand loadInkFileCommand;
        private RelayCommand saveInkFileCommand;
        private RelayCommand transformTextAndShapesCommand;
        private RelayCommand clearAllCommand;

        private bool enableTouch = true;
        private bool enableMouse = true;
        private bool enablePen = true;

        private bool enableLassoSelection;

        public InkSmartCanvasViewViewModel()
        {
        }

        public void Initialize(
            InkStrokesService strokeService,
            InkLassoSelectionService lassoSelectionService,
            InkNodeSelectionService nodeSelectionService,
            InkPointerDeviceService pointerDeviceService,
            InkUndoRedoService undoRedoService,
            InkTransformService transformService,
            InkFileService fileService)
        {
            _strokeService = strokeService;
            _lassoSelectionService = lassoSelectionService;
            _nodeSelectionService = nodeSelectionService;
            _pointerDeviceService = pointerDeviceService;
            _undoRedoService = undoRedoService;
            _transformService = transformService;
            _fileService = fileService;

            _pointerDeviceService.DetectPenEvent += (s, e) => EnableTouch = false;
        }

        public RelayCommand UndoCommand => undoCommand
           ?? (undoCommand = new RelayCommand(() =>
           {
               ClearSelection();
               _undoRedoService.Undo();
           }));

        public RelayCommand RedoCommand => redoCommand
           ?? (redoCommand = new RelayCommand(() =>
           {
               ClearSelection();
               _undoRedoService.Redo();
           }));

        public RelayCommand LoadInkFileCommand => loadInkFileCommand
           ?? (loadInkFileCommand = new RelayCommand(async () =>
           {
               ClearSelection();
               var fileLoaded = await _fileService.LoadInkAsync();

               if (fileLoaded)
               {
                   _transformService.ClearTextAndShapes();
                   _undoRedoService.Reset();
               }
           }));

        public RelayCommand SaveInkFileCommand => saveInkFileCommand
           ?? (saveInkFileCommand = new RelayCommand(async () =>
           {
               ClearSelection();
               await _fileService.SaveInkAsync();
           }));

        public RelayCommand TransformTextAndShapesCommand => transformTextAndShapesCommand
           ?? (transformTextAndShapesCommand = new RelayCommand(async () =>
           {
               var result = await _transformService.TransformTextAndShapesAsync();
               if (result.TextAndShapes.Any())
               {
                   ClearSelection();
                   _undoRedoService.AddOperation(new TransformUndoRedoOperation(result, _strokeService));
               }
           }));

        public RelayCommand ClearAllCommand => clearAllCommand
           ?? (clearAllCommand = new RelayCommand(ClearAll));

        public bool EnableTouch
        {
            get => enableTouch;
            set
            {
                Param_Setter(ref enableTouch, value);
                _pointerDeviceService.EnableTouch = value;
            }
        }

        public bool EnableMouse
        {
            get => enableMouse;
            set
            {
                Param_Setter(ref enableMouse, value);
                _pointerDeviceService.EnableMouse = value;
            }
        }

        public bool EnablePen
        {
            get => enablePen;
            set
            {
                Param_Setter(ref enablePen, value);
                _pointerDeviceService.EnablePen = value;
            }
        }

        public bool EnableLassoSelection
        {
            get => enableLassoSelection;
            set
            {
                Param_Setter(ref enableLassoSelection, value);
                ConfigLassoSelection(value);
            }
        }

        private void ConfigLassoSelection(bool enableLasso)
        {
            if (enableLasso)
            {
                _lassoSelectionService.StartLassoSelectionConfig();
            }
            else
            {
                _lassoSelectionService.EndLassoSelectionConfig();
            }
        }

        private void ClearSelection()
        {
            _nodeSelectionService.ClearSelection();
            _lassoSelectionService.ClearSelection();
        }

        private void ClearAll()
        {
            ClearSelection();
            _strokeService.ClearStrokes();
            _transformService.ClearTextAndShapes();
            _undoRedoService.Reset();
        }
    }
}

File: templates/Uwp/Pages/InkSmartCanvasVB/ViewModels/InkSmartCanvasViewViewModel.vb
Imports System.Windows.Input
Imports Param_ItemNamespace.Services.Ink
Imports Param_ItemNamespace.Services.Ink.UndoRedo

Namespace ViewModels
    Public Class InkSmartCanvasViewViewModel
        Inherits System.ComponentModel.INotifyPropertyChanged

        Private _strokeService As InkStrokesService
        Private _lassoSelectionService As InkLassoSelectionService
        Private _nodeSelectionService As InkNodeSelectionService
        Private _pointerDeviceService As InkPointerDeviceService
        Private _undoRedoService As InkUndoRedoService
        Private _transformService As InkTransformService
        Private _fileService As InkFileService
        Private _undoCommand As ICommand
        Private _redoCommand As ICommand
        Private _loadInkFileCommand As ICommand
        Private _saveInkFileCommand As ICommand
        Private _transformTextAndShapesCommand As ICommand
        Private _clearAllCommand As ICommand
        Private _enableTouch As Boolean = True
        Private _enableMouse As Boolean = True
        Private _enablePen As Boolean = True
        Private _enableLassoSelection As Boolean

        Public Sub New()
        End Sub

        Public Sub Initialize(strokeService As InkStrokesService, lassoSelectionService As InkLassoSelectionService, nodeSelectionService As InkNodeSelectionService, pointerDeviceService As InkPointerDeviceService, undoRedoService As InkUndoRedoService, transformService As InkTransformService, fileService As InkFileService)
            _strokeService = strokeService
            _lassoSelectionService = lassoSelectionService
            _nodeSelectionService = nodeSelectionService
            _pointerDeviceService = pointerDeviceService
            _undoRedoService = undoRedoService
            _transformService = transformService
            _fileService = fileService
            AddHandler _pointerDeviceService.DetectPenEvent, Sub(s, e) EnableTouch = False
        End Sub

        Public ReadOnly Property UndoCommand As ICommand
            Get
                If _undoCommand Is Nothing Then
                    _undoCommand = New RelayCommand(Sub()
                                                        ClearSelection()
                                                        _undoRedoService?.Undo()
                                                    End Sub)
                End If

                Return _undoCommand
            End Get
        End Property

        Public ReadOnly Property RedoCommand As ICommand
            Get
                If _redoCommand Is Nothing Then
                    _redoCommand = New RelayCommand(Sub()
                                                        ClearSelection()
                                                        _undoRedoService?.Redo()
                                                    End Sub)
                End If

                Return _redoCommand
            End Get
        End Property

        Public ReadOnly Property LoadInkFileCommand As ICommand
            Get
                If _loadInkFileCommand Is Nothing Then
                    _loadInkFileCommand = New RelayCommand(Async Sub()
                                                               ClearSelection()
                                                               Dim fileLoaded = Await _fileService?.LoadInkAsync()

                                                               If fileLoaded Then
                                                                   _transformService.ClearTextAndShapes()
                                                                   _undoRedoService?.Reset()
                                                               End If
                                                           End Sub)
                End If

                Return _loadInkFileCommand
            End Get
        End Property

        Public ReadOnly Property SaveInkFileCommand As ICommand
            Get
                If _saveInkFileCommand Is Nothing Then
                    _saveInkFileCommand = New RelayCommand(Async Sub()
                                                               ClearSelection()
                                                               Await _fileService?.SaveInkAsync()
                                                           End Sub)
                End If

                Return _saveInkFileCommand
            End Get
        End Property

        Public ReadOnly Property TransformTextAndShapesCommand As ICommand
            Get
                If _transformTextAndShapesCommand Is Nothing Then
                    _transformTextAndShapesCommand = New RelayCommand(Async Sub()
                                                                          Dim result = Await _transformService.TransformTextAndShapesAsync()

                                                                          If result.TextAndShapes.Any() Then
                                                                              ClearSelection()
                                                                              _undoRedoService.AddOperation(New TransformUndoRedoOperation(result, _strokeService))
                                                                          End If
                                                                      End Sub)
                End If

                Return _transformTextAndShapesCommand
            End Get
        End Property

        Public ReadOnly Property ClearAllCommand As ICommand
            Get
                If _clearAllCommand Is Nothing Then
                    _clearAllCommand = New RelayCommand(AddressOf ClearAll)
                End If

                Return _clearAllCommand
            End Get
        End Property

        Public Property EnableTouch As Boolean
            Get
                Return _enableTouch
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableTouch, value)
                _pointerDeviceService.EnableTouch = value
            End Set
        End Property

        Public Property EnableMouse As Boolean
            Get
                Return _enableMouse
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableMouse, value)
                _pointerDeviceService.EnableMouse = value
            End Set
        End Property

        Public Property EnablePen As Boolean
            Get
                Return _enablePen
            End Get
            Set(value As Boolean)
                [Param_Setter](_enablePen, value)
                _pointerDeviceService.EnablePen = value
            End Set
        End Property

        Public Property EnableLassoSelection As Boolean
            Get
                Return _enableLassoSelection
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableLassoSelection, value)
                ConfigLassoSelection(value)
            End Set
        End Property

        Private Sub ConfigLassoSelection(enableLasso As Boolean)
            If enableLasso Then
                _lassoSelectionService.StartLassoSelectionConfig()
            Else
                _lassoSelectionService.EndLassoSelectionConfig()
            End If
        End Sub

        Private Sub ClearSelection()
            _nodeSelectionService.ClearSelection()
            _lassoSelectionService.ClearSelection()
        End Sub

        Private Sub ClearAll()
            ClearSelection()
            _strokeService.ClearStrokes()
            _transformService.ClearTextAndShapes()
            _undoRedoService.Reset()
        End Sub
    End Class
End Namespace
